### PR TITLE
Small usearch index improvements: metrics and configurable internal data type

### PR DESCRIPTION
**Pull request by al13n321**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

The metrics should be very useful for figuring out performance, and f16 make the index smaller and maybe faster.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit a4113ef8acb7e6d94d0ed170d42b043b7e9d41b4 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54103/a4113ef8acb7e6d94d0ed170d42b043b7e9d41b4/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/MergeTree/MergeTreeIndexUSearch.cpp [hunk 1]
```diff
@@ -165,12 +177,16 @@ void MergeTreeIndexAggregatorUSearch<Metric>::update(const Block & block, size_t
         if (!index->reserve(unum::usearch::ceil2(index->size() + num_rows)))
             throw Exception(ErrorCodes::CANNOT_ALLOCATE_MEMORY, "Could not reserve memory for usearch index");
 
-        if (auto rc = index->add(index->size(), array.data()); !rc)
-            throw Exception(ErrorCodes::INCORRECT_DATA, rc.error.release());
-        for (size_t current_row = 1; current_row < num_rows; ++current_row)
-            if (auto rc = index->add(index->size(), &array[offsets[current_row - 1]]); !rc)
+        for (size_t current_row = 0; current_row < num_rows; ++current_row)
+        {
+            auto rc = index->add(static_cast<uint32_t>(index->size()), &array[offsets[current_row - 1]]);
```
Comment (rschu1ze) [inline src/Storages/MergeTree/MergeTreeIndexUSearch.cpp:182]: If I see correctly, one of these overloads is called:
```cpp
add_result_t add(key_t key, b1x8_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_b1x8); }
add_result_t add(key_t key, i8_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_i8); }
add_result_t add(key_t key, f16_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_f16); }
add_result_t add(key_t key, f32_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_f32); }
add_result_t add(key_t key, f64_t const* vector, std::size_t thread = any_thread(), bool force_vector_copy = true) { return add_(key, vector, thread, force_vector_copy, casts_.from_f64); }
```
It looks like we need to convert each input value from single-precision float to half-precision float first (something like [this](https://stackoverflow.com/questions/1659440/32-bit-to-16-bit-floating-point-conversion))?

Comment (rschu1ze) [reply src/Storages/MergeTree/MergeTreeIndexUSearch.cpp:182]: Oh, it seems they do the casting inside `add()` by themselves.

**Comment by al13n321** [comment]
Made the f16/f32/whatever configurable with a second argument to the index, like `usearch('L2Distance', 'f16')`. Set default back to f32 because it's faster (but bigger).

File: src/Storages/MergeTree/MergeTreeIndexUSearch.cpp (hunk 1 above)
Comment (ashvardanian) [reply src/Storages/MergeTree/MergeTreeIndexUSearch.cpp:206]: Yes, we do. Type-punning is one of the core purposes of the `index_dense_gt` layer.

**Comment by ashvardanian** [comment]
@al13n321, this would heavily depend on the hardware support and is evident in brute-force scenarios, [as discussed in this thread](https://github.com/unum-cloud/usearch/issues/249#issuecomment-1706594975). The recent release brings some great optimizations for x86 CPUs with `f16c` and `avx512fp16` support. I've also switched from GCC builtins to inline assembly to support both GCC and Clang, as [Clang is currently lacking those `__builtin`-s](https://github.com/llvm/llvm-project/issues/65320).

**Comment by al13n321** [comment]
Changed the default to f16 now that there's dynamic dispatch for the AVX512 implementation.

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
(I accidentally squashed some smaller fixes of mine with @al13n321 's commit, sorry - had to force-push. But the PR looks good now, let's get it in.)

**Pull request merged**